[TorchScript] Support user defined classes as constants

bwasti opened the pull request:
Summary: User defined classes can be used as constants.  This is useful when freezing and removing the module from the graph.

Test Plan: waitforsadcastle

Differential Revision: D23994974

bzinodev commented on aten/src/ATen/core/ivalue.cpp:
```diff
@@ -459,6 +459,9 @@ std::ostream& IValue::repr(
       return out << enum_holder->qualifiedClassName() << "." <<
           enum_holder->name();
     }
+    case IValue::Tag::Object: {
```
Should this be TORCH_CHECK?  we don't expect to handle this case in the future.

eellison reviewed (commented):
I don't think we should merge this if it's going to silently break the ability of a user to save and load their model in combination with torch.jit.freeze. 

However, I think it would be very easy to add serialization support in combination with https://github.com/pytorch/pytorch/pull/37631, which is accepted but not merged.

bzinodev reviewed (changes requested):
Where the tests?

bzinodev commented on torch/csrc/jit/serialization/python_print.cpp:
```diff
@@ -864,6 +864,7 @@ struct PythonPrintImpl {
 
   void printConstant(TaggedStringStream& stmt, const IValue& v) {
     const auto customFormatter = [&](std::ostream& ss, const IValue& v) {
+      // TODO Add support for serialization of custom classes
```
Should there be an assertion here? 
To make sure serialization is not invoked where there is a custom class as a constant

ZolotukhinM commented on torch/csrc/jit/serialization/python_print.cpp:
```diff
@@ -864,7 +864,8 @@ struct PythonPrintImpl {
     const auto customFormatter = [&](std::ostream& ss, const IValue& v) {
-      if (v.isTensor() || containsNonASCIIString(v)) {
+      if (v.isTensor() || containsNonASCIIString(v) ||
+          (v.isObject() && !v.type()->is_module())) {
```
Should we also handle Function objects here?

bwasti replied:
not sure, why would we?

eellison replied:
Yea, don't think that's needed. Could consider using https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/core/ivalue.cpp#L125 to check for contained object constants, although i suspect that's somewhat unlikely.

ZolotukhinM replied:
By handling I actually meant not handling it 😀, but bailing out if we see them (similarly to how bail out for modules here).

ZolotukhinM commented on torch/csrc/jit/passes/freeze_module.cpp:
```diff
@@ -77,7 +77,7 @@ class AttributePropagator {
     };
     auto applyOptimizations = [](std::shared_ptr<Graph>& subgraph) {
-      runOptimization(subgraph, /* unroll? */ false);
+      runOptimization(subgraph, /* unroll? */ false, /* const_prop? */ false);
```
The comment "const_prop?" is misleading: we're running const prop anyway.

ZolotukhinM commented on aten/src/ATen/core/ivalue.cpp:
```diff
@@ -504,6 +504,9 @@ std::ostream& IValue::repr(
           enum_holder->name();
     }
+    case IValue::Tag::Object: {
+      TORCH_INTERNAL_ASSERT(false, "repr() not defined on: ", v.tagKind(), ". Perhaps you've frozen a module with custom classes?");
```
Nit: the line seems too long.

bwasti commented on torch/csrc/jit/passes/constant_propagation.cpp:
```diff
@@ -25,6 +27,18 @@ c10::optional<std::vector<IValue>> runNodeIfInputsAreConstant(const Node* n) {
+  // Weak form of const propagation
+  if (ignore_custom_classes) {
+    for (auto o : n->outputs()) {
+      if (auto ival = toIValue(o)) {
```
this is a way to check if its constant (returns nullptr if not).  seemed to be the style of the extent code, so I copied it.

eellison replied:
+ 1 to checking Type here

bwasti replied:
this is more involved than seems obvious, I'd need to change the interface to custom class checks to allow non-ivalue checks

https://codebrowser.bddppq.com/pytorch/pytorch/aten/src/ATen/core/custom_class.cpp.html#_ZN5torch13isCustomClassERKN3c106IValueE

bwasti replied:
since custom_class is in ATen core, adding things from ir.h is super hard IIRC (requires a refactor)

eellison commented on torch/csrc/jit/passes/xnnpack_rewrite.cpp:
```diff
@@ -348,6 +352,9 @@ void FoldPrePackingOps(script::Module& m) {
   PrePackingOpsFolder(m, filter_fn, "prepack_folding");
+  auto graph = m.get_method("forward").graph();
+  // Folding requires a const propagation through user defined classes
+  ConstantPropagation(graph, false);
```
IMO,  wanting to restrict behavior of constant prop is kind of niche behavior & not wanted by most users, so the  flag as it stands is  preferable.

jerryzh168 commented:
Thanks @bwasti if this is enabled by default, you can also remove https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/passes/quantization/finalize.cpp#L99 and see if anything breaks (the pass is added specifically to fold constant objects)

bwasti committed:
[TorchScript] Support user defined classes as constants (#5062)

Summary:
Pull Request resolved: https://github.com/pytorch/glow/pull/5062

Pull Request resolved: https://github.com/pytorch/pytorch/pull/45556

User defined classes can be used as constants.  This is useful when freezing and removing the module from the graph.

Test Plan: waitforsadcastle

Reviewed By: eellison

Differential Revision: D23994974

fbshipit-source-id: 6c494269e222b7e1b5ecddf0c460ae8c09ac0556
